### PR TITLE
Improve curvilinear geometries support for vector field conversion and plot modifications

### LINKED ISSUE #2096: Potential curvilinear geometry support for plot modifications?
Some plot modifications such as `annotate_line_integral_convolution` and `annotate_streamlines` potentially works fine with curvilinear geometries, by converting vector fields from curvilinear to cartesian geometry. An example as follows:

```python
import yt
import numpy as np

@yt.derived_field(name="magnetic_cyl_to_x", units="gauss")
def _magnetic_cyl_to_x(field,data):
    return data["magnetic_field_r"] * np.cos(data["theta"]) - data["magnetic_field_theta"] * np.sin(data["theta"])

@yt.derived_field(name="magnetic_cyl_to_y", units="gauss")
def _magnetic_cyl_to_y(field,data):
    return data["magnetic_field_r"] * np.sin(data["theta"]) + data["magnetic_field_theta"] * np.cos(data["theta"])

filename = 'taylor_hdf5_plt_cnt_0150'
ds = yt.load(filename) # load 3D cylindrical data
ds.coordinates.name = "cartesian" # cheat here to avoid unsupported geometry error
slc = yt.SlicePlot(ds, "z", "magnetic_field_strength")
slc.annotate_line_integral_convolution("magnetic_cyl_to_x", "magnetic_cyl_to_y")
slc.save("lic.png")
slc.annotate_clear()
slc.annotate_streamlines("magnetic_cyl_to_x", "magnetic_cyl_to_y")
slc.save("stream.png")
```

And it will produce the following figures:

lic
stream

The streamline annotation does not work properly because `yt.utilities.lib.pixelization_routines.pixelize_cartesian` is used; but if we use `plot.data.ds.coordinates.pixelize`, the streamline will work. So the question is why the former routine instead of the latter is chosen for pixelization?

### LINKED ISSUE #2103: Velocity vectors do not get converted to cylindrical components in 3D cylindrical geo
### Bug report

**Bug summary**

In 3d cylindrical coordinates, velocity vectors do not get converted to (r, theta, z) components (but magnetic field vectors do).

**Code for reproduction**

Datafile: [https://www.dropbox.com/s/jmstyqittx8ubz2/cyl3d.hdf5?dl=0](https://www.dropbox.com/s/jmstyqittx8ubz2/cyl3d.hdf5?dl=0)

```python
import yt
  
ds = yt.load("cyl3d.hdf5")
print(ds.derived_field_list)
print(ds.all_data()["velocity_magnitude"][0])
```

### LINKED ISSUE #2113: covering_grid returns wrong coordinate values for cylindrical geometry
### Bug report

**Bug summary**

`covering_grid` returns wrong coordinate values for cylindrical geometry.

**Code for reproduction**

Datafile: [https://www.dropbox.com/s/jmstyqittx8ubz2/cyl3d.hdf5?dl=0](https://www.dropbox.com/s/jmstyqittx8ubz2/cyl3d.hdf5?dl=0)

```python
import yt
  
ds = yt.load("cyl3d.hdf5")
grid = ds.covering_grid(0, ds.domain_left_edge, ds.domain_dimensions)
print("r", grid["r"].min(), grid["r"].max())
print("z", grid["z"].min(), grid["z"].max())
print("theta", grid["theta"].min(), grid["theta"].max())
```

**Actual outcome**

```
yt : [INFO     ] 2018-12-05 20:33:32,776 Particle file found: cyl3d.hdf5
yt : [INFO     ] 2018-12-05 20:33:32,820 Parameters: current_time              = 0.0
yt : [INFO     ] 2018-12-05 20:33:32,820 Parameters: domain_dimensions         = [32 32 32]
yt : [INFO     ] 2018-12-05 20:33:32,820 Parameters: domain_left_edge          = [ 1.  -0.5  0. ]
yt : [INFO     ] 2018-12-05 20:33:32,821 Parameters: domain_right_edge         = [2.         0.5        6.28318531]
yt : [INFO     ] 2018-12-05 20:33:32,821 Parameters: cosmological_simulation   = 0.0
r 1.015625 cm 1.984375 cm
z 0.09817477042468103 cm 6.1850105367549055 cm
theta 0.09817477042468103 dimensionless 6.1850105367549055 dimensionless
```

The line of `z` coordinate is wrong, which gives the same numbers with `theta`.

### LINKED ISSUE #2116: Select region gives YTFieldNotFound error for 2D cylindrical dataset
### Bug report

**Bug summary**

Select region gives YTFieldNotFound error for 2D cylindrical dataset.

**Code for reproduction**

```python
import yt

ds = yt.load('WDMerger_hdf5_chk_1000.hdf5')
ad = ds.all_data()
upper_ad = ad.cut_region(["obj['z'] > 0 “])

sp = ds.sphere((0, 0, 0), (6E10, 'cm'), data_source=upper_ad)
sp.quantities.total_mass()
```

**Actual outcome**

```
YTFieldNotFound                           Traceback (most recent call last)
<ipython-input-10-833a8830b022> in <module>()
      4 
      5 sp = ds.sphere((0, 0, 0), (6E10, 'cm'), data_source=upper_ad)
----> 6 sp.quantities.total_mass()

/global/u2/c/ccfeng/yt/yt/data_objects/derived_quantities.py in __call__(self)
    206         fi = self.data_source.ds.field_info
    207         if ("gas", "cell_mass") in fi:
--> 208             gas = super(TotalMass, self).__call__([('gas', 'cell_mass')])
    209         else:
    210             gas = self.data_source.ds.arr([0], 'g')

/global/u2/c/ccfeng/yt/yt/data_objects/derived_quantities.py in __call__(self, fields)
    176     def __call__(self, fields):
    177         fields = ensure_list(fields)
--> 178         rv = super(TotalQuantity, self).__call__(fields)
    179         if len(rv) == 1: rv = rv[0]
    180         return rv

/global/u2/c/ccfeng/yt/yt/data_objects/derived_quantities.py in __call__(self, *args, **kwargs)
     66         chunks = self.data_source.chunks([], chunking_style="io")
     67         storage = {}
---> 68         for sto, ds in parallel_objects(chunks, -1, storage = storage):
     69             sto.result = self.process_chunk(ds, *args, **kwargs)
     70         # Now storage will have everything, and will be done via pickling, so

/global/u2/c/ccfeng/yt/yt/utilities/parallel_tools/parallel_analysis_interface.py in parallel_objects(objects, njobs, storage, barrier, dynamic)
    505     # this will prevent intermediate objects from being created.
    506     oiter = itertools.islice(enumerate(objects), my_new_id, None, njobs)
--> 507     for result_id, obj in oiter:
    508         if storage is not None:
    509             rstore = ResultsStorage()

/global/u2/c/ccfeng/yt/yt/data_objects/data_containers.py in chunks(self, fields, chunking_style, **kwargs)
   1261     def chunks(self, fields, chunking_style, **kwargs):
   1262         # This is an iterator that will yield the necessary chunks.
-> 1263         self.get_data() # Ensure we have built ourselves
   1264         if fields is None: fields = []
   1265         # chunk_ind can be supplied in the keyword arguments.  If it's a

/global/u2/c/ccfeng/yt/yt/data_objects/data_containers.py in get_data(self, fields)
   1304     def get_data(self, fields=None):
   1305         if self._current_chunk is None:
-> 1306             self.index._identify_base_chunk(self)
   1307         if fields is None: return
   1308         nfields = []

/global/u2/c/ccfeng/yt/yt/geometry/grid_geometry_handler.py in _identify_base_chunk(self, dobj)
    287             dobj._chunk_info[0] = weakref.proxy(dobj)
    288         elif getattr(dobj, "_grids", None) is None:
--> 289             gi = dobj.selector.select_grids(self.grid_left_edge,
    290                                             self.grid_right_edge,
    291                                             self.grid_levels)

/global/u2/c/ccfeng/yt/yt/data_objects/data_containers.py in selector(self)
   1254 
   1255         if self._data_source is not None:
-> 1256             self._selector = compose_selector(self, self._data_source.selector, sclass(self))
   1257         else:
   1258             self._selector = sclass(self)

/global/u2/c/ccfeng/yt/yt/data_objects/data_containers.py in selector(self)
   1254 
   1255         if self._data_source is not None:
-> 1256             self._selector = compose_selector(self, self._data_source.selector, sclass(self))
   1257         else:
   1258             self._selector = sclass(self)

/global/u2/c/ccfeng/yt/yt/geometry/selection_routines.pyx in yt.geometry.selection_routines.CutRegionSelector.__init__()

/global/u2/c/ccfeng/yt/yt/data_objects/data_containers.py in __getitem__(self, key)
    246         Returns a single field.  Will add if necessary.
    247         """
--> 248         f = self._determine_fields([key])[0]
    249         if f not in self.field_data and key not in self.field_data:
    250             if f in self._container_fields:

/global/u2/c/ccfeng/yt/yt/data_objects/data_containers.py in _determine_fields(self, fields)
   1157                     raise YTFieldNotParseable(field)
   1158                 ftype, fname = field
-> 1159                 finfo = self.ds._get_field_info(ftype, fname)
   1160             elif isinstance(field, DerivedField):
   1161                 ftype, fname = field.name

/global/u2/c/ccfeng/yt/yt/data_objects/static_output.py in _get_field_info(self, ftype, fname)
    796                     self._last_finfo = self.field_info[(ftype, fname)]
    797                     return self._last_finfo
--> 798         raise YTFieldNotFound((ftype, fname), self)
    799 
    800     def _setup_classes(self):

YTFieldNotFound: Could not find field '('index', 'x')' in WDMerger_hdf5_chk_1000.hdf5.
```

### PR DESCRIPTION
**Pull request by jisuoqing**
## PR Summary

- Fix the bug that in curvilinear geometries, fluid vector field aliases (except B fields) do not get converted to curvilinear coordinates, fix #2103 (many particle fields are hard-coded and to fix them is beyond this PR)
- Disable incorrect vector field operations in curvilinear geometries (to correct all operations such as divergence and curl for curvilinear geometries is time consuming and beyond this PR)
- Enable some plot modifications (annotate quivers, streamlines, line integral convolutions, magnetic fields and velocities) for 2D polar and 3D cylindrical geometries, by adding the vector fields projected to Cartesian x-y plane `vector_field_cartesian_(xyz)`, close #2096
- Fix the bug of wrong coordinate values returned by `covering_grid`, close #2113.
- Fix bug in CutRegionSelector for curvilinear geometry, partially fix #2116 (only fix the first test script; the second one with particle file does not work yet, which should be dealt with in a separate PR...)
- Fix wrong plot window width for 3D cylindrical geometry when `r > 0`.

## PR Checklist

- [x] Code passes flake8 checker
- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Looks like there are some flake8 issues, a few answer test failures (please look closely at those, if they're pixel-level differences just bump the answers, please let me know if you need a hand with the answer tests), and a few unit test failures.

Overall this looks really great, thank you for working on this! Please let me know if you need a hand with the tests, it looks like you're breaking a couple of the answer tests and introducing tiny pixel-level differences. I wouldn't worry too much about keeping the answers exactly the same, just bump the answers if all the answer test failures look like single pixel-level differences.

**Comment by jisuoqing** [comment]
Thanks a lot for kindly offering your help! Let me see what I can fix first and will definitely ask for your help when needed.

**Comment by jisuoqing** [comment]
@ngoldbaum Ok -- now I met something I do not understand: it seems that lots of image differences are caused by the dashed vs. solid lines...

File: yt/visualization/plot_modifications.py [hunk 1]
```diff
@@ -405,15 +423,13 @@ class QuiverCallback(PlotCallback):
     substantial variation in field strength.
     """
     _type_name = "quiver"
-    _supported_geometries = ("cartesian", "spectral_cube", "cylindrical-2d")
+    _supported_geometries = ("cartesian", "spectral_cube", "polar", "cylindrical")
     def __init__(self, field_x, field_y, factor=16, scale=None,
-                 scale_units=None, normalize=False, bv_x=0, bv_y=0,
```
Comment (ngoldbaum) [inline yt/visualization/plot_modifications.py]: Why did you remove these two keyword arguments? Since they're exposed in the user-facing API I think we need to keep them.

**Comment by ngoldbaum** [comment]
Hmm so at least one of those failures is real. To reproduce, see the following test script:

```
import yt
ds = yt.load("WindTunnel/windtunnel_4lev_hdf5_plt_cnt_0030")
plot = yt.SlicePlot(ds, 'z', 'density')
plot.set_buff_size((600, 800))
plot.annotate_velocity()
plot.save()
```

Before your PR, this script produces this image:

https://i.imgur.com/tA1p23y.png

After it produces this image:

https://i.imgur.com/c3pIebD.png

I'm not sure what's up with the dashed/solid line differences on the contour plots in the test failures on python3, I can't reproduce that locally. I'd say try fixing the real failure I found above then if the python3 failures keep on happening, we can bump the answer numbers.

See also the comment I left inline about the removed keyword arguments, I think we still need to support those as they're in the public API.

PS it looks like you did this PR on the master branch. In the future you want to avoid that (by making a new branch before commiting for a pull request) since it will make it difficult for you to simultaneously work on two pull requests, or for you to pull from upstream's master branch if you need to do that in your development repository for some reason.

File: yt/visualization/plot_modifications.py (hunk 1 above)
Comment (jisuoqing) [reply yt/visualization/plot_modifications.py]: This is because `plot.data.ds.coordinates.pixelize` does not accept `plot.data[self.field_x]`, but only `self.field_x`, so I can not subtract `bv_x` from the field `self.field._x`. Or you have any idea?

Comment (ngoldbaum) [reply yt/visualization/plot_modifications.py]: You could add those as optional new arguments to `pixelize`?

Comment (jisuoqing) [reply yt/visualization/plot_modifications.py]: Sure! I'll take a look at it.

**Comment by jisuoqing** [comment]
> plot.set_buff_size((600, 800))

Interesting... it seems that this line broke the image -- I'll investigate this further.

**Comment by jisuoqing** [comment]
@ngoldbaum I try to use a new dataset for test script, but it can not be read in -- do you have any idea why? Thanks!

**Comment by jisuoqing** [comment]
@yt-fido test this please

**Comment by jisuoqing** [comment]
no problem! I'm traveling anyway.

File: yt/geometry/coordinates/cylindrical_coordinates.py [hunk 2]
```diff
@@ -211,7 +211,7 @@ def sanitize_width(self, axis, width, depth):
             width = [2.0 * np.pi * self.ds.domain_width.uq,
                      self.ds.domain_width[z_ax]]
         elif name == "theta":
-            width = [self.ds.domain_right_edge[r_ax],
+            width = [self.ds.domain_width[r_ax],
```
[the remark this answers is not in the document]
Comment (jisuoqing) [reply yt/geometry/coordinates/cylindrical_coordinates.py]: This is because for cylindrical geometry, Rmin is not always zero. When Rmin is larger than zero, the old code will give a larger width and create empty space.

**Comment by jisuoqing** [comment]
Since I'm attending a workshop, updating this PR is a bit lagged. Most of your comments are dealt with except for the docs. I'm not sure where to mention it -- in plot modifications? If we attached one curvilinear example to each updated callbacks, it seems redundant.

**Comment by ngoldbaum** [comment]
Maybe we need a general section on dealing with data in curvilinear geometries along with caveats about what is and isn't currently supported? If there isn't a natural place to document these new features I'm not going to block merging this on creating that new doc section, but it would be nice to create an issue about it so we can take care of it before the next release.

@MatthewTurk it would be great if you could look this over as you've thought more carefully about supporting curvilinear data than anyone else.

**Comment by MatthewTurk** [comment]
@ngoldbaum @jisuoqing sorry for not getting to this sooner.

@jisuoqing Would you be able to have a call to go over this sometime either tomorrow (Friday) or early next week?

**Comment by jisuoqing** [comment]
@MatthewTurk Thanks! However I'm traveling abroad so it's a bit hard for me to set a time. I'll be back by the end of Jan and will let you know then :)

**Comment by jisuoqing** [comment]
@MatthewTurk Ok, I got back recently and am happy to do a call -- preferably any afternoon.

File: yt/geometry/coordinates/cylindrical_coordinates.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatthewTurk) [reply yt/geometry/coordinates/cylindrical_coordinates.py]: That's fair, yeah.  I am optimistic that with changes like those from @munkm that help with the different projections we can make this a bit easier to work with.

File: yt/fields/fluid_fields.py [hunk 3]
```diff
@@ -188,6 +191,9 @@ def _mean_molecular_weight(field, data):
                           weight="cell_mass")
 
 def setup_gradient_fields(registry, grad_field, field_units, slice_info = None):
+    # Current implementation for gradient is not valid for curvilinear geometries
+    if is_curvilinear(registry.ds.geometry): return
+
```
Comment (jzuhone) [inline yt/fields/fluid_fields.py]: Maybe we need to log a warning here?

Comment (jzuhone) [reply yt/fields/fluid_fields.py]: Ah, never mind. I see why we wouldn't want to do this.

**Comment by jzuhone** [comment]
@jisuoqing, have you looked at the code in `yt/fields/magnetic_fields.py` to see if you need to do anything there?

**Comment by jisuoqing** [comment]
@jzuhone thanks for reminding! I've looked at the `magnetic_field.py` and it is good to me. One of the original issues related to this PR is that the magnetic fields can be handled correctly, but not the velocity fields.
